### PR TITLE
feature(fir): controlled reward/value categorical representation

### PR DESCRIPTION
**Pull request by Firerozes**
Current implementation doesn't allow flexibility on categorical representations and is flawed. Therefore this is both a fix and an enhancement.

## How it was before

- `config` arguments `support_scale` only allowed categorical representations between `-support_scale` and `support_scale` (included)
- `config` arguments `support_scale` was used both for value AND reward representation
- there was some redundance between `config` arguments `support_scale` ; `reward_support_size` and `value_support_size` : we were compelled to have `reward_support_size` = `value_support_size` = 2*`support_scale`+1 (and I guess if the user made a mistake on that messed up stuff could be happening)
- `phi_transform` and `DiscreteSupport` from `lzero/policy/scaling_transform.py` were implemented to theorically support `delta` != 1. ; although never used in practice. Though, `phi_transform` wasn't correct for the case `delta` != 1.

## MAIN COMPONENTS
- replaced `config` arguments `support_scale` ; `reward_support_size` and `value_support_size` by `config.policy.model.reward_support_range` and `config.policy.model.value_support_range`, allowing the user to specifically select how he wants both to be represented
- adapted `lzero/policy/scaling_transform.py` functions to work conveniently with this new way of doing, fix undesired behaviours, and optimize a bit
- adapted every `lzero/policy`, every `lzero/mcts/tree_search` and every `lzero/mcts/buffer` plus differentiating the `self.value_support` from the `self.reward_support`
- adapted every `lzero/policy` and every `lzero/mcts/tree_search` plus differentiating the `self.value_inverse_scalar_transform_handle` from the `self.reward_inverse_scalar_transform_handle`
- adapted every `lzero/mcts/buffer` plus differentiating usage of `inverse_scalar_transform` between values and rewards
- adapted evey `lzero/model`, giving them default config `reward_support_range` and `value_support_range` and initializing in their `__init__` `reward_support_size` and `value_support_size` accordingly
- adapted every comment and description accordingly

## MINOR COMPONENTS
- adapted accordingly EVERY `docs/source/tutorials/algos` (`docs/source/tutorials/config/config_zh.md` needs a small review from you since I don't speak Chinese 🙂)
- adapted accordingly EVERY `docs/source/tutorials/config`
- adapted accordingly EVERY agent config for EVERY env
- adapted `zoo/game_2048/config/stochastic_muzero_2048_config.py` : now it is possible to truly replicate Stochastic MuZero paper : "We used 601 bins for both the value and the reward predictions with both the value and the reward being able to represent values between [0, 600.]" (source : https://openreview.net/pdf?id=X6D9bAHhBQ1 page 14)
- adapted accordingly EVERY `lzero/mcts/tests`

### BONUS
- created a `self.mse_loss = torch.nn.MSELoss()` in `lzero/policy/stochastic_muzero.py` because before it was initialized at each call and that's not clean

I tested with a Stochastic MuZero 2048 training run and everything works as a charm.
I stay available if any detail is needed.

### CODE REVIEW

File: lzero/policy/scaling_transform.py [hunk 1]
```diff
@@ -151,19 +145,21 @@ def phi_transform(discrete_support: DiscreteSupport, x: torch.Tensor) -> torch.T
...
+    min_bound = arange[0, 0]
+    max_bound = arange[0, -1]
+    size = discrete_support.size
+    step = discrete_support.step
+
+    x.clamp_(min_bound, max_bound)
+    x_low_idx = ((x - min_bound)/step).floor().long()
+    x_high_idx = x_low_idx + 1
+    x_high_idx = x_high_idx.clamp(0, size - 1)
+    x_low_val = torch.take_along_dim(arange, x_low_idx, dim=1)
```
Comment (puyuan1996) [inline lzero/policy/scaling_transform.py]: Why  do we need `take_along_dim` here？A more optimized approach might be like:

```
def phi_transform(
    discrete_support: DiscreteSupport,
    x: torch.Tensor,
) -> torch.Tensor:
    """
    Overview:
        Map a real-valued scalar to a categorical distribution over a discrete support using linear interpolation (a.k.a. “soft” one-hot).

        For each scalar value the probability mass is split between the two
        nearest support atoms so that their weighted sum equals the original
        value (MuZero, Appendix F).

    Arguments:
        - discrete_support : DiscreteSupport
            Container with the support values (must be evenly spaced).
        - x : torch.Tensor
            Input tensor of arbitrary shape ``(...,)`` containing real numbers.

    Returns:
        - torch.Tensor
            Tensor of shape ``(*x.shape, N)`` where ``N = discrete_support.size``.
            The last dimension is a probability distribution (sums to 1).

    Notes
    -----
    • No in-place ops on the input are used, improving autograd safety.  
    • Only one `scatter_add_` kernel is launched for efficiency.  
    """
    # --- constants ----------------------------------------------------------
    min_bound = discrete_support.arange[0]
    max_bound = discrete_support.arange[-1]
    step      = discrete_support.step
    size      = discrete_support.size

    # --- 1. clip to the valid range ----------------------------------------
    x = x.clamp(min_bound, max_bound)

    # --- 2. locate neighbouring indices ------------------------------------
    pos       = (x - min_bound) / step          # continuous position
    low_idx   = torch.floor(pos).long()         # lower index
    high_idx  = low_idx + 1                     # upper index (may overflow)

    # --- 3. linear interpolation weights -----------------------------------
    p_high = (pos - low_idx.float())            # distance to lower atom
    valid_high = (high_idx < size).float()      # mask for overflow
    p_high = p_high * valid_high
    p_low  = 1.0 - p_high                       # complementary mass

    # --- 4. stack indices / probs and scatter ------------------------------
    idx   = torch.stack([low_idx,
                         torch.clamp(high_idx, max=size - 1)], dim=-1)  # (*x, 2)
    prob  = torch.stack([p_low, p_high], dim=-1)                        # (*x, 2)

    target = torch.zeros(*x.shape, size,
                         dtype=x.dtype, device=x.device)

    target.scatter_add_(-1, idx.unsqueeze(-1), prob.unsqueeze(-1))
    return target
```

Before merging the PR, we need to ensure that the optimized version of `phi_transform` is functionally consistent with the original version while exhibiting improved performance. This can be validated using synthetic data for unit testing. Thank you for your contribution.

File: lzero/policy/sampled_unizero.py [hunk 2]
```diff
@@ -60,9 +60,10 @@ class SampledUniZeroPolicy(UniZeroPolicy):
             num_res_blocks=1,
             # (int) The number of channels of hidden states in MuZero model.
             num_channels=64,
-            # (int) The scale of supports used in categorical distribution.
-            # This variable is only effective when ``categorical_distribution=True``.
-            support_scale=50,
+            # (tuple) The range of supports used in categorical distribution.
+            # These variables are only effective when ``model.categorical_distribution=True``.
+            reward_support_range=(-50., 51., 1.),
```
Comment (puyuan1996) [inline lzero/policy/sampled_unizero.py:65]: Hello, before merging the PR, please ensure that all parts using scaling_transform are updated accordingly. For example, some policy/config modules (such as unizero/dmc) still need to be adapted. Thank you very much for your effort and support!

File: lzero/policy/scaling_transform.py [hunk 3]
```diff
@@ -151,19 +145,21 @@ def phi_transform(discrete_support: DiscreteSupport, x: torch.Tensor) -> torch.T
...
-    max = discrete_support.max
-    set_size = discrete_support.set_size
-    delta = discrete_support.delta
-
-    x.clamp_(min, max)
-    x_low = x.floor()
-    x_high = x.ceil()
-    p_high = x - x_low
+    arange = discrete_support.arange
+    min_bound = arange[0, 0]
```
Comment (puyuan1996) [inline lzero/policy/scaling_transform.py]: why arange[0, 0] rather than arange[0]?

Comment (E-Delabrouille) [reply lzero/policy/scaling_transform.py]: Because in `DiscreteSupport` I decided to put `self.arange = torch.arange(start, stop, step).unsqueeze(0).to(device)` rather than `self.arange = np.arange(start, stop, step)`, so that `self.arange` is of shape `(1, size)` and not `(size)`, because it's the shape they're needed in in both `inverse_scalar_transform` and `InverseScalarTransform`. Therefore the tensor is directly stored as it's needed. Thererefore to retrieve its min_bound the correct call is indeed `arange[0, 0]` rather than `arange[0]`

File: lzero/policy/scaling_transform.py (hunk 1 above)
Comment (E-Delabrouille) [reply lzero/policy/scaling_transform.py]: I made today a new commit on categorical_representation branch, merging the qualities of both your approach of `phi_transform` and mine. Since `p_high` is calculated from `low_idx`, its values are always correct and need no `valid_high` approval. Therefore these two lines are not needed :
```python
    valid_high = (high_idx < size).float()      # mask for overflow
    p_high = p_high * valid_high
```

Also added a tiny typing correction elsewhere in the commit, stuff that slipped under my attention on the initial commit.

You say "we need to ensure that the optimized version of `phi_transform` is functionally consistent with the original version while exhibiting improved performance" :
-> I tested on a small Jupyter Notebook and everything works as intended. Though, note that this new version is **not** conceived as an optimization. It could work faster, or slower than before, this is not my point. The old version was **incorrect for each delta != 1**. Therefore, this update is needed from a **correctness** point of view - which takes priority over performance. Especially now that I give the possibility to create discrete supports with steps != 1.

File: lzero/policy/sampled_unizero.py (hunk 2 above)
Comment (E-Delabrouille) [reply lzero/policy/sampled_unizero.py:65]: Well, I did a (very) extensive search and updated everything that I saw...
I updated every policy, model, config, tree_search, buffer, test, tutorial file that was using the categorical representations...that is, every file where there was a reference to it, I modified it. The files with no reference to it used the default configuration...and still will after the PR if we don't modify them.

Is there something I missed ?

**Comment by puyuan1996** [comment]
Hello, thank you very much for your effort and contribution. Upon review, I can confirm that you have properly adapted all sections related to categorical representations. Additionally, you have successfully verified the correctness of the rewritten scaling_transform.py locally. Based on these factors, this PR is ready for merging. Once again, thank you for your excellent work.
